### PR TITLE
NativeAnimatedModule TrackingAnimatedNode

### LINKED ISSUE #2940: Implement Tracking AnimatedNodes
The NativeAnimatedModule does not have an implementation for the Tracking AniamtedNodeType.

### PR DESCRIPTION
**Pull request by StephenLPeters**
This pull request has a fully implemented tracking node for the NativeANimatedModule.  The animation matches the JS conterparts for frame and spring based animations and decay animations cannot track values by design.

Fixes #2940

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/3041)

### CODE REVIEW

File: vnext/ReactUWP/Modules/Animated/NativeAnimatedNodeManager.cpp [hunk 1]
```diff
@@ -145,6 +148,94 @@ void NativeAnimatedNodeManager::StopAnimation(int64_t animationId) {
...
+void NativeAnimatedNodeManager::StartTrackingAnimatedNode(
+    int64_t animationId,
+    int64_t animatedNodeTag,
+    int64_t animatedToValueTag,
+    const folly::dynamic &animationConfig,
+    const Callback &endCallback,
+    const std::shared_ptr<NativeAnimatedNodeManager> &manager,
+    bool track) {
+  auto updatedAnimationConfig = animationConfig;
+  for (auto iterator = m_activeAnimations.begin();
```
Comment (chrisglein) [inline vnext/ReactUWP/Modules/Animated/NativeAnimatedNodeManager.cpp]: Can this just be a ranged for?
`for (auto &item : m_activeAnimations)`

File: vnext/ReactUWP/Modules/Animated/NativeAnimatedNodeManager.cpp [hunk 2]
```diff
@@ -145,6 +148,94 @@ void NativeAnimatedNodeManager::StopAnimation(int64_t animationId) {
...
+        case AnimationType::Frames:
+          updatedAnimationConfig.insert(
+              static_cast<folly::StringPiece>(s_framesName),
+              [animationConfig, iterator]() {
+                auto frames = folly::dynamic::array();
+                for (auto const &frame :
+                     animationConfig.find("frames").dereference().second) {
+                  frames.push_back(0.0);
+                }
+                for (auto const &frame : iterator->second->Frames()) {
```
Comment (chrisglein) [inline vnext/ReactUWP/Modules/Animated/NativeAnimatedNodeManager.cpp]: To append you can do `frames.insert(frames.end(), other.begin(), other.end())` (where other is iterator->second)

File: vnext/ReactUWP/Modules/Animated/SpringAnimationDriver.h [hunk 3]
```diff
@@ -16,44 +16,45 @@ class SpringAnimationDriver : public AnimationDriver {
...
+double m_displacementFromRestThreshold{0};
+bool m_overshootClampingEnabled{0};
+int m_iterations{0};
+folly::dynamic m_dynamicToValues{};
+
+static constexpr std::string_view s_springStiffnessParameterName{"stiffness"};
+static constexpr std::string_view s_springDampingParameterName{"damping"};
+static constexpr std::string_view s_springMassParameterName{"mass"};
+static constexpr std::string_view s_initialVelocityParameterName{
+    "initialVelocity"};
```
Comment (chrisglein) [inline vnext/ReactUWP/Modules/Animated/SpringAnimationDriver.h]: I think pretty universally everyone thinks the ColumnLimit of 80 in the .clang-format is too low. We snapped someone else's as a starting point. If you want to take a PR to bump that to even just 100 I think you'd be welcomed with open arms :)

File: vnext/ReactUWP/Modules/Animated/TrackingAnimatedNode.cpp [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+void TrackingAnimatedNode::StartAnimation() {
+  if (auto const strongManager = m_manager.lock()) {
+    if (auto const toValueNode =
+            strongManager->GetValueAnimatedNode(m_toValueId)) {
+      toValueNode->AddActiveTrackingNode(m_tag);
+      m_animationConfig.insert(
+          static_cast<folly::StringPiece>(s_toValueIdName),
+          toValueNode->Value());
+      //m_animationConfig.at("frames").insert()
```
Comment (chrisglein) [inline vnext/ReactUWP/Modules/Animated/TrackingAnimatedNode.cpp]: What's this?

File: vnext/ReactUWP/Modules/Animated/NativeAnimatedNodeManager.cpp (hunk 2 above)
Comment (StephenLPeters) [reply vnext/ReactUWP/Modules/Animated/NativeAnimatedNodeManager.cpp]: Folly::Dynamic::array does not have an iterator based insert function.

File: vnext/ReactUWP/Modules/Animated/ValueAnimatedNode.cpp [hunk 5]
```diff
@@ -40,7 +40,10 @@ double ValueAnimatedNode::RawValue() {
 }
 
 void ValueAnimatedNode::RawValue(double value) {
-  m_propertySet.InsertScalar(s_valueName, static_cast<float>(value));
+  if (RawValue() != value) {
```
Comment (licanhua) [inline vnext/ReactUWP/Modules/Animated/ValueAnimatedNode.cpp]: >(RawValue() != value [](start = 5, length = 20)

For double, maybe it's not the best way to do the comparison. I don't know if you need to switch to [https://bitbashing.io/comparing-floats.html](https://bitbashing.io/comparing-floats.html)

File: vnext/ReactUWP/Modules/Animated/AnimationDriver.cpp [hunk 6]
```diff
@@ -36,42 +36,45 @@ AnimationDriver::~AnimationDriver() {
...
   }
   scopedBatch.End();
 
-  m_scopedBatchCompletedToken = scopedBatch.Completed(
-      [endCallback = m_endCallback, animatedValue, id = m_id](
-          auto sender, auto) {
-        if (endCallback) {
-          endCallback(std::vector<folly::dynamic>{
+  m_scopedBatchCompletedToken =
+      scopedBatch.Completed([this](auto sender, auto) {
```
Comment (chrisglein) [inline vnext/ReactUWP/Modules/Animated/AnimationDriver.cpp]: This event token is being saved, but I don't see anything that's protecting the [this] capture. The event could happen after the lifetime of this object, correct?

File: vnext/ReactUWP/Modules/Animated/NativeAnimatedNodeManager.cpp [hunk 7]
```diff
@@ -145,6 +148,95 @@ void NativeAnimatedNodeManager::StopAnimation(int64_t animationId) {
...
+      animation->StopAnimation(true);
+      m_activeAnimations.erase(animationId);
+      StartTrackingAnimatedNode(
+          animationId,
+          animatedValueTag,
+          animatedToValueTag,
+          animationConfig,
+          endCallback,
+          manager,
+          false);
```
Comment (chrisglein) [inline vnext/ReactUWP/Modules/Animated/NativeAnimatedNodeManager.cpp]: comment bool parameter

File: vnext/ReactUWP/Modules/Animated/NativeAnimatedNodeManager.cpp [hunk 8]
```diff
@@ -145,6 +148,95 @@ void NativeAnimatedNodeManager::StopAnimation(int64_t animationId) {
...
+        default:
+          assert(false);
+          break;
+      }
+
+      break;
+    }
+  }
+  if (track)
+    m_trackingAndLeadNodeTags.push_back(
```
Comment (chrisglein) [inline vnext/ReactUWP/Modules/Animated/NativeAnimatedNodeManager.cpp]: braces around the if statement

File: vnext/ReactUWP/Modules/Animated/NativeAnimatedNodeManager.cpp [hunk 9]
```diff
@@ -351,5 +455,13 @@ TransformAnimatedNode *NativeAnimatedNodeManager::GetTransformAnimatedNode(
   }
   return static_cast<TransformAnimatedNode *>(nullptr);
 }
+
+TrackingAnimatedNode *NativeAnimatedNodeManager::GetTrackingAnimatedNode(
+    int64_t tag) {
+  if (m_trackingNodes.count(tag)) {
+    return m_trackingNodes.at(tag).get();
+  }
+  return static_cast<TrackingAnimatedNode *>(nullptr);
```
Comment (chrisglein) [inline vnext/ReactUWP/Modules/Animated/NativeAnimatedNodeManager.cpp]: Was this cast not implicit? Why do you need it?

File: vnext/ReactUWP/Modules/Animated/SpringAnimationDriver.cpp [hunk 10]
```diff
@@ -93,6 +96,15 @@ SpringAnimationDriver::MakeAnimation(const folly::dynamic &config) {
 std::tuple<float, double> SpringAnimationDriver::GetValueAndVelocityForTime(
     double time,
     double startValue) {
+  const auto toValue = [this, time, startValue]() {
+    const auto frameFromTime = static_cast<int>(time * 60.0);
+    if (frameFromTime < static_cast<int>(m_dynamicToValues.size())) {
+      return startValue +
+          (m_dynamicToValues[frameFromTime].asDouble() *
+           (m_endValue - startValue));
```
Comment (chrisglein) [inline vnext/ReactUWP/Modules/Animated/SpringAnimationDriver.cpp]:
```suggestion
          (m_endValue - startValue));
```

File: vnext/ReactUWP/Modules/Animated/SpringAnimationDriver.h [hunk 11]
```diff
@@ -16,19 +16,20 @@ class SpringAnimationDriver : public AnimationDriver {
...
  private:
   std::tuple<float, double> GetValueAndVelocityForTime(
       double time,
       double startValue);
-  bool IsAtRest(
-      double currentVelocity,
-      double currentPosition,
-      double startPosition);
+  bool
+  IsAtRest(double currentVelocity, double currentPosition, double endValue);
```
Comment (chrisglein) [inline vnext/ReactUWP/Modules/Animated/SpringAnimationDriver.h]: What was wrong about how it was formatted before?

Comment (StephenLPeters) [reply vnext/ReactUWP/Modules/Animated/SpringAnimationDriver.h]: clang changed it with the name change <shrug>

---
In reply to: [324896812](https://github.com/microsoft/react-native-windows/pull/3041#discussion_r324896812) [](ancestors = 324896812)

File: vnext/ReactUWP/Modules/Animated/NativeAnimatedNodeManager.cpp (hunk 9 above)
Comment (StephenLPeters) [reply vnext/ReactUWP/Modules/Animated/NativeAnimatedNodeManager.cpp]: I think i copied this from a lambda definition where it is needed, but in this definition it is not. thanks!

---
In reply to: [324895296](https://github.com/microsoft/react-native-windows/pull/3041#discussion_r324895296) [](ancestors = 324895296)

**Pull request merged**